File: AdvantageFramework.Core/GoogleServices/TokenService.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using System.Text;
using System.Linq;

namespace AdvantageFramework.Core.GoogleServices
{
    public partial class TokenService
    {

        /* TODO ERROR: Skipped RegionDirectiveTrivia */


        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */


        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private static Assembly _Assembly = null;
        private static MethodInfo[] _MethodInfos = null;

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */


        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */


        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private static Assembly GetServicesAssembly()
        {
            Assembly GetServicesAssemblyRet = default;
            try
            {
                if (_Assembly is null)
                {
                    _Assembly = Assembly.Load("AdvantageFramework.GoogleServices");
                }
            }
            catch (Exception ex)
            {
                _Assembly = null;
            }

            GetServicesAssemblyRet = _Assembly;
            return GetServicesAssemblyRet;
        }

        private static MethodInfo[] GetServiceMethodInfos()
        {
            MethodInfo[] GetServiceMethodInfosRet = default;
            Assembly Assembly = null;
            Type BaseMethodType = null;
            try
            {
                if (_MethodInfos is null)
                {
                    Assembly = GetServicesAssembly();
                    if (Assembly is object)
                    {
                        BaseMethodType = LoadBaseMethodTypeShared(Assembly);
                        if (BaseMethodType is object)
                        {
                            _MethodInfos = BaseMethodType.GetMethods();
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                _MethodInfos = null;
            }

            GetServiceMethodInfosRet = _MethodInfos;
            return GetServiceMethodInfosRet;
        }

        private static Type LoadBaseMethodTypeShared(Assembly Assembly)
        {
            Type LoadBaseMethodTypeSharedRet = default;

            // objects
            Type BaseMethodType = null;
            try
            {
                BaseMethodType = Assembly.GetTypes().Where(Type => Type.FullName == "AdvantageFramework.GoogleServices.Methods").SingleOrDefault();
            }
            catch (Exception ex)
            {
                BaseMethodType = null;
            }
            finally
            {
                LoadBaseMethodTypeSharedRet = BaseMethodType;
            }

            return LoadBaseMethodTypeSharedRet;
        }

        public static string GetOrRefreshToken(string ConnectionString, bool UseWindowsAuthentication, string EmployeeCode, string[] Scopes, string ClientCode)
        {
            string GetOrRefreshTokenRet = default;

            // objects
            string AccessToken = string.Empty;
            MethodInfo[] MethodInfos = null;
            try
            {
                MethodInfos = GetServiceMethodInfos();
                if (MethodInfos is object)
                {
                    AccessToken = (MethodInfos.Where(Method => Method.Name == "GetOrRefreshToken").SingleOrDefault().Invoke(null, new[] { ConnectionString, (object)UseWindowsAuthentication, EmployeeCode, Scopes, ClientCode })).ToString();
                }
            }
            catch (Exception)
            {
                AccessToken = null;
            }
            finally
            {
                GetOrRefreshTokenRet = AccessToken;
            }

            return GetOrRefreshTokenRet;
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
    }
}

File: AdvantageFramework/Google Services/TokenService.vb
Namespace GoogleServices

    Public Class TokenService

#Region " Constants "



#End Region

#Region " Enum "



#End Region

#Region " Variables "

        Private Shared _Assembly As System.Reflection.Assembly = Nothing
        Private Shared _MethodInfos As System.Reflection.MethodInfo() = Nothing

#End Region

#Region " Properties "



#End Region

#Region " Methods "



#Region " Shared "

        Private Shared Function GetServicesAssembly() As System.Reflection.Assembly

            Try

                If _Assembly Is Nothing Then

                    _Assembly = System.Reflection.Assembly.Load("AdvantageFramework.GoogleServices")

                End If

            Catch ex As Exception
                _Assembly = Nothing
            End Try

            GetServicesAssembly = _Assembly

        End Function
        Private Shared Function GetServiceMethodInfos() As System.Reflection.MethodInfo()

            Dim Assembly As System.Reflection.Assembly = Nothing
            Dim BaseMethodType As System.Type = Nothing

            Try

                If _MethodInfos Is Nothing Then

                    Assembly = GetServicesAssembly()

                    If Assembly IsNot Nothing Then

                        BaseMethodType = LoadBaseMethodTypeShared(Assembly)

                        If BaseMethodType IsNot Nothing Then

                            _MethodInfos = BaseMethodType.GetMethods()

                        End If

                    End If

                End If

            Catch ex As Exception
                _MethodInfos = Nothing
            End Try

            GetServiceMethodInfos = _MethodInfos

        End Function
        Private Shared Function LoadBaseMethodTypeShared(Assembly As System.Reflection.Assembly) As System.Type

            'objects
            Dim BaseMethodType As System.Type = Nothing

            Try

                BaseMethodType = Assembly.GetTypes().Where(Function([Type]) [Type].FullName = "AdvantageFramework.GoogleServices.Methods").SingleOrDefault

            Catch ex As Exception
                BaseMethodType = Nothing
            Finally
                LoadBaseMethodTypeShared = BaseMethodType
            End Try

        End Function
        Public Shared Function GetOrRefreshToken(ConnectionString As String, UseWindowsAuthentication As Boolean, EmployeeCode As String, Scopes() As String, ClientCode As String) As String

            'objects
            Dim AccessToken As String = String.Empty
            Dim MethodInfos As System.Reflection.MethodInfo() = Nothing

            Try

                MethodInfos = GetServiceMethodInfos()

                If MethodInfos IsNot Nothing Then

                    AccessToken = MethodInfos.Where(Function(Method) Method.Name = "GetOrRefreshToken").SingleOrDefault.Invoke(Nothing, {ConnectionString, UseWindowsAuthentication, EmployeeCode, Scopes, ClientCode})

                End If

            Catch ex As Exception
                AccessToken = Nothing
            Finally
                GetOrRefreshToken = AccessToken
            End Try

        End Function

#End Region

#End Region

    End Class

End Namespace

